website/test/test-demo.mjs

```diff
@@ -51,76 +96,192 @@ describe("markEditorText", () => {
-    markEditorText(editor, dom.window, [{ begin: 7, end: 7 + "three".length }]);
-    assert.strictEqual(editor.innerHTML, "one<br>two<mark>three</mark>");
+    markEditorText(editor, dom.window, [
+      {
+        begin: 7,
+        end: 7 + "three".length,
+        message: "unknow error",
+        code: unKnownErrorCode,
+        severity: 5,
+      },
+    ]);
+    assert.strictEqual(
+      editor.innerHTML,
+      'one<br>two<mark data-message="unknow error" data-code="E0XXX" data-severity="5">three</mark>'
+    );
   });
 
   it("marking preserves <br> after inserted mark", () => {
     let editor = preElementWithHTML("onetwo<br>three");
-    markEditorText(editor, dom.window, [{ begin: 0, end: 3 }]);
-    assert.strictEqual(editor.innerHTML, "<mark>one</mark>two<br>three");
+    markEditorText(editor, dom.window, [
+      {
+        begin: 0,
+        end: 3,
+        message: "unknow error",
+        code: unKnownErrorCode,
+        severity: 5,
+      },
+    ]);
+    assert.strictEqual(
+      editor.innerHTML,
+      '<mark data-message="unknow error" data-code="E0XXX" data-severity="5">one</mark>two<br>three'
+    );
   });
 
   it("mark exactly over existing <mark>", () => {
     let editor = preElementWithHTML("<mark>hello</mark> world");
-    markEditorText(editor, dom.window, [{ begin: 0, end: 5 }]);
-    assert.strictEqual(editor.innerHTML, "<mark>hello</mark> world");
+    markEditorText(editor, dom.window, [
+      {
+        begin: 0,
+        end: 5,
+        message: "unknow error",
+        code: unKnownErrorCode,
+        severity: 5,
+      },
+    ]);
+    assert.strictEqual(
+      editor.innerHTML,
+      '<mark data-message="unknow error" data-code="E0XXX" data-severity="5">hello</mark> world'
+    );
   });
 
   it("mark starts at end of existing <mark>", () => {
     let editor = preElementWithHTML("<mark>hello</mark>world");
-    markEditorText(editor, dom.window, [{ begin: 5, end: 10 }]);
-    assert.strictEqual(editor.innerHTML, "hello<mark>world</mark>");
+    markEditorText(editor, dom.window, [
+      {
+        begin: 5,
+        end: 10,
+        message: "unknow error",
+        code: unKnownErrorCode,
+        severity: 5,
+      },
+    ]);
+    assert.strictEqual(
+      editor.innerHTML,
+      'hello<mark data-message="unknow error" data-code="E0XXX" data-severity="5">world</mark>'
+    );
   });
 
   it("add empty mark", () => {
     let editor = preElementWithHTML("helloworld");
-    markEditorText(editor, dom.window, [{ begin: 5, end: 5 }]);
-    assert.strictEqual(editor.innerHTML, "hello<mark></mark>world");
+    markEditorText(editor, dom.window, [
+      {
+        begin: 5,
+        end: 5,
+        message: "unknow error",
+        code: unKnownErrorCode,
+        severity: 5,
+      },
+    ]);
+    assert.strictEqual(
+      editor.innerHTML,
+      'hello<mark data-message="unknow error" data-code="E0XXX" data-severity="5"></mark>world'
+    );
```

Comment:
Hmm, these tests seem fragile. What if the order of attributes changes? What if the attribute order is randomized?

Most of these tests are concerned with the *position* of the marks. We shouldn't be testing unimportant *attributes* of the marks.

An exception is tests which have multiple marks. However, your commit doesn't distinguish the marks, so having the tests check the attributes is just noise which makes it harder to understand the tests.

I see two reasonable solutions:

* Don't give message/code/severity to `markEditorText`, and have `markEditorText` omit the `data-` attributes.
* In each assertion, filter the attributes out of the HTML. We could do this with ugly string hacks, or we could do this by removing attributes with DOM APIs.

Reply:
I create a condition to check if there is message/code/severity before create  data-attributes